### PR TITLE
Move inline css and js to external files for SwaggerUI and ReDoc

### LINKED ISSUE #2881: [Feature request]: Remove inline code in Swagger UI index.html
### Is your feature request related to a specific problem? Or an existing feature?

Not really a problem, but I noticed something:
2 years ago Swagger UI removed their inline CSS and JavaScript from the index.html file. The current file is [here](https://github.com/swagger-api/swagger-ui/blob/master/dist/index.html), the reason for the change can be found in swagger-api/swagger-ui#7295

It would be nice if Swashbuckle.AspNetCore.SwaggerUI would also do it this way.

### Describe the solution you'd like

The inline CSS can simply be replaced by a link to the index.css file, just like the original index.html does it now. This also has the benefit that future updates in the index.css file are available in Swashbuckle.AspNetCore.SwaggerUI automatically.

The inline JavaScript code in Swashbuckle.AspNetCore.SwaggerUI is heavily modified compared to the original code (now in the file swagger-initializer.js). The modified code could be extracted to a custom js file, but then `SwaggerUIMiddleware` would need to be changed to process that file instead of index.html.
Note: This might need additional consideration concerning Browser-side caching.

### Additional context

The same might be true for Redoc UI, I haven't checked that.

### PR DESCRIPTION
**Pull request by junior-santana**
# Pull Request

## The issue or feature being addressed

Resolves #2881.

## Details on the issue fix or feature implementation

Moved inline javascript to separate files
ReDoc: Moved inline CSS to separate file
SwaggerUi: Reused CSS file from the package

Middleware changes to respond with corresponding files

### CODE REVIEW

File: test/Swashbuckle.AspNetCore.IntegrationTests/ReDocIntegrationTests.cs [hunk 1]
```diff
@@ -19,6 +19,18 @@ public async Task RoutePrefix_RedirectsToIndexUrl()
             Assert.Equal("api-docs/index.html", response.Headers.Location.ToString());
         }
 
+        [Fact]
+        public async Task RedocMiddleware_ReturnsInitializerScript()
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/ReDocIntegrationTests.cs:23]: Missing a test for the CSS file?

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.html [hunk 2]
```diff
@@ -5,113 +5,17 @@
...
-    </style>
-    %(HeadContent)
+    %(HeadContent)   
 </head>
 
 <body>
     <div id="swagger-ui"></div>
 
-    <!-- Workaround for https://github.com/swagger-api/swagger-editor/issues/1371 -->
-    <script>
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/index.html:36]: Is there a possibility that the ordering here was important, so combining the scripts might cause an issue?

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 3]
```diff
@@ -76,17 +94,17 @@ public async Task IndexUrl_ReturnsInterceptors_IfConfigured()
...
 
             Assert.Contains("\"RequestInterceptorFunction\":", content);
             Assert.Contains("\"ResponseInterceptorFunction\":", content);
         }
 
         [Theory]
-        [InlineData("/swagger/index.html", new [] { "Version 1.0", "Version 2.0" })]
-        [InlineData("/swagger/1.0/index.html", new [] { "Version 1.0" })]
-        [InlineData("/swagger/2.0/index.html", new [] { "Version 2.0" })]
+        [InlineData("/swagger/index.js", new[] { "Version 1.0", "Version 2.0" })]
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:105]: Similar question here.

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 4]
```diff
@@ -26,25 +26,43 @@ public class SwaggerUIIntegrationTests
...
+            Type startupType,
+            string indexJsPath)
+        {
+            var client = new TestSite(startupType).BuildClient();
+
+            var indexResponse = await client.GetAsync(indexJsPath);
+            Assert.Equal(HttpStatusCode.OK, indexResponse.StatusCode);
+
+            var indexContent = await indexResponse.Content.ReadAsStringAsync();
+            Assert.Contains("SwaggerUIBundle", indexContent);
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:41]: Need to assert the templating is working.

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs [hunk 5]
```diff
@@ -68,8 +68,16 @@ public async Task Invoke(HttpContext httpContext)
 
             var isGet = HttpMethods.IsGet(httpMethod);
 
+            var match = Regex.Match(path, $"^/{Regex.Escape(_options.RoutePrefix)}/?(index.html|index.js)$", RegexOptions.IgnoreCase);
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs:71]: Refactor to remove the duplication with `index.(html|js)`?

File: src/Swashbuckle.AspNetCore.SwaggerUI/Swashbuckle.AspNetCore.SwaggerUI.csproj [hunk 6]
```diff
@@ -16,7 +16,10 @@
 
   <ItemGroup>
     <EmbeddedResource Include="node_modules/swagger-ui-dist/**/*" Exclude="**/*/index.html;**/*/*.map;**/*/*.json;**/*/*.md;**/*/swagger-ui-es-*" />
+    <None Remove="fix-edge-fetch.js" />
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/Swashbuckle.AspNetCore.SwaggerUI.csproj:19]: Does this file exist?

File: src/Swashbuckle.AspNetCore.ReDoc/index.js [hunk 7]
```diff
@@ -0,0 +1 @@
+Redoc.init('%(SpecUrl)', JSON.parse('%(ConfigObject)'), document.getElementById('redoc-container'));
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.ReDoc/index.js:1]: There don't appear to have been any code changes that replace these templated vars.

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.html [hunk 8]
```diff
@@ -5,113 +5,17 @@
...
-            const ui = SwaggerUIBundle(configObject);
-
-            ui.initOAuth(oauthConfigObject);
-
-            // End Swagger UI call region
-
-            window.ui = ui
-        }
-    </script>
+    <script src="index.js"></script>
```
Comment (cremor) [inline src/Swashbuckle.AspNetCore.SwaggerUI/index.html:19]: I don't know if it makes a difference, but the [original index.html file](https://github.com/swagger-api/swagger-ui/blob/v5.17.14/dist/index.html) defines a `charset="UTF-8"` attribute on all 3 script tags.

Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.SwaggerUI/index.html:19]: I played around with some diacritics/ideograms (Portuguese, Cyrillic, Japanese) and couldn't find any issue but it's worth adding it for consistency. I guess I should define it in the header in the middleware as well

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs (hunk 5 above)
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs:71]: ✔️

File: src/Swashbuckle.AspNetCore.SwaggerUI/Swashbuckle.AspNetCore.SwaggerUI.csproj [hunk 9]
```diff
@@ -6,7 +6,7 @@
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <NoWarn>$(NoWarn);1591</NoWarn>
     <PackageTags>swagger;documentation;discovery;help;webapi;aspnet;aspnetcore</PackageTags>
-    <SignAssembly>true</SignAssembly>
+    <SignAssembly>true</SignAssembly>
```
[the remark this answers is not in the document]
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.SwaggerUI/Swashbuckle.AspNetCore.SwaggerUI.csproj:9]: ✔️

File: src/Swashbuckle.AspNetCore.ReDoc/ReDocMiddleware.cs [hunk 10]
```diff
@@ -67,9 +67,11 @@ public async Task Invoke(HttpContext httpContext)
                 return;
             }
 
-            if (httpMethod == "GET" && Regex.IsMatch(path, $"/{_options.RoutePrefix}/?index.html",  RegexOptions.IgnoreCase))
+            var match = Regex.Match(path, $"^/{Regex.Escape(_options.RoutePrefix)}/?(index.html|index.css|index.js)$", RegexOptions.IgnoreCase);
```
[the remark this answers is not in the document]
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.ReDoc/ReDocMiddleware.cs:70]: ✔️

[the remark this answers is not in the document]
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.ReDoc/ReDocMiddleware.cs:70]: ✔️

File: src/Swashbuckle.AspNetCore.ReDoc/index.js (hunk 7 above)
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.ReDoc/index.js:1]: There have been some changes in the `RespondWithFile` method, the replace itself works regardless

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.html (hunk 2 above)
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.SwaggerUI/index.html:36]: Edit: if this is still required then the order is essential and I will likely have to go with the extra js file.
Idk when it was fixed on Edge (It's fine in current version - v126.0.2592.87)

File: src/Swashbuckle.AspNetCore.SwaggerUI/Swashbuckle.AspNetCore.SwaggerUI.csproj [hunk 11]
```diff
@@ -16,7 +16,9 @@
 
   <ItemGroup>
     <EmbeddedResource Include="node_modules/swagger-ui-dist/**/*" Exclude="**/*/index.html;**/*/*.map;**/*/*.json;**/*/*.md;**/*/swagger-ui-es-*" />
+    <None Remove="index.js" />
     <EmbeddedResource Include="index.html" />
+    <EmbeddedResource Include="index.js" />
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/Swashbuckle.AspNetCore.SwaggerUI.csproj:21]:
```suggestion
    <None Remove="index.html;index.js" />
    <EmbeddedResource Include="index.html;index.js" />
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.html (hunk 2 above)
Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerUI/index.html:36]: I guess we can ditch it if it appears to be vestigial and then we can add it back if it turns out it's still a problem for people.

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs [hunk 12]
```diff
@@ -110,23 +114,36 @@ private static void RespondWithRedirect(HttpResponse response, string location)
...
-            response.ContentType = "text/html;charset=utf-8";
 
-            using (var stream = _options.IndexStream())
+            Stream stream;
+
+            if (fileName == "index.js")
+            {
+                response.ContentType = "application/javascript;charset=utf-8";
+                stream = typeof(SwaggerUIMiddleware).GetTypeInfo().Assembly
+                            .GetManifestResourceStream($"Swashbuckle.AspNetCore.SwaggerUI.{fileName}");
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIMiddleware.cs:127]: Similar here.

File: src/Swashbuckle.AspNetCore.ReDoc/ReDocMiddleware.cs [hunk 13]
```diff
@@ -97,21 +103,40 @@ private void RespondWithRedirect(HttpResponse response, string location)
...
+                case "index.js":
+                    response.ContentType = "application/javascript;charset=utf-8";
+                    stream = typeof(ReDocMiddleware).GetTypeInfo().Assembly
+                                .GetManifestResourceStream($"Swashbuckle.AspNetCore.ReDoc.{fileName}");
+                    break;
+                default:
+                    response.ContentType = "text/html;charset=utf-8";
+                    stream = _options.IndexStream();
+                    break;
+            }
```
[the remark this answers is not in the document]
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.ReDoc/ReDocMiddleware.cs]: Done

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.html (hunk 2 above)
Comment (junior-santana) [reply src/Swashbuckle.AspNetCore.SwaggerUI/index.html:36]: Agreed.
Done.

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    if (!configObject.hasOwnProperty("validatorUrl"))
+        configObject.validatorUrl = null
+
+    // If oauth2RedirectUrl isn't specified, use the built-in default
+    if (!configObject.hasOwnProperty("oauth2RedirectUrl"))
+        configObject.oauth2RedirectUrl = (new URL("oauth2-redirect.html", window.location.href)).href;
+
+    // Apply mandatory parameters
+    configObject.dom_id = "#swagger-ui";
+    configObject.presets = [SwaggerUIBundle.presets.apis, SwaggerUIStandalonePreset];
```
Comment (cremor) [inline src/Swashbuckle.AspNetCore.SwaggerUI/index.js:50]: The [default config file](https://github.com/swagger-api/swagger-ui/blob/v5.17.14/dist/swagger-initializer.js) also sets `plugins: [SwaggerUIBundle.plugins.DownloadUrl]`. I don't know what that does, but maybe it should be set here too?

File: src/Swashbuckle.AspNetCore.ReDoc/ResourceHelper.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿using System.IO;
+using System.Reflection;
+
+namespace Swashbuckle.AspNetCore.ReDoc;
+
+internal class ResourceHelper
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.ReDoc/ResourceHelper.cs:6]:
```suggestion
internal static class ResourceHelper
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/ResourceHelper.cs [hunk 16]
```diff
@@ -0,0 +1,13 @@
+﻿using System.IO;
+using System.Reflection;
+
+namespace Swashbuckle.AspNetCore.SwaggerUI;
+
+internal class ResourceHelper
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/ResourceHelper.cs:6]:
```suggestion
internal static class ResourceHelper
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.js (hunk 14 above)
Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerUI/index.js:50]: I suggest we leave it as is, particularly if we don't know why it isn't there already.

**Review by martincostello — APPROVED** [review]

**Pull request merged**